File: battle_engine/app/contracts/consumers/turn_actions_contract.rb
# frozen_string_literal: true

module Contracts
  module Consumers
    class TurnActionsContract < Dry::Validation::Contract
      params do
        required(:battle_id).filled(:string)
        required(:turn).filled(:integer)
        optional(:timestamp).maybe(:string)
        required(:actions).array(:hash) do
          required(:action).filled(:string, included_in?: %w[attack attack_switch switch])
          required(:player_id).filled(:integer)
          optional(:id).maybe(:integer)
          optional(:pokemon_id).maybe(:integer)
          optional(:pokemon_switched_id).maybe(:integer)
          optional(:targets).maybe { array? & each(:string) }
        end
      end

      rule(:actions) do
        next unless values[:actions].is_a?(Array)

        values[:actions].each_with_index do |action_item, index|
          case action_item[:action]
          when 'attack'
            key([:actions, index, :id]).failure('must be filled for attack action') if action_item[:id].nil?
            key([:actions, index, :pokemon_id]).failure('must be filled for attack action') if action_item[:pokemon_id].nil?
            if action_item[:targets].nil? || !action_item[:targets].is_a?(Array) || action_item[:targets].empty?
              key([:actions, index, :targets]).failure('must be filled for attack action')
            end
          when 'attack_switch'
            key([:actions, index, :id]).failure('must be filled for attack_switch action') if action_item[:id].nil?
            key([:actions, index, :pokemon_id]).failure('must be filled for attack_switch action') if action_item[:pokemon_id].nil?
            key([:actions, index, :pokemon_switched_id]).failure('must be filled for attack_switch action') if action_item[:pokemon_switched_id].nil?
            if action_item[:targets].nil? || !action_item[:targets].is_a?(Array) || action_item[:targets].empty?
              key([:actions, index, :targets]).failure('must be filled for attack_switch action')
            end
          when 'switch'
            key([:actions, index, :pokemon_id]).failure('must be filled for switch action') if action_item[:pokemon_id].nil?
          end
        end
      end
    end
  end
end

File: battle_real_time/lib/battle_real_time/contracts/publishers/battle_actions/turn_actions_contract.ex
defmodule BattleRealTime.Contracts.Publishers.BattleActions.TurnActionsContract do
  @moduledoc """
  Validates the consolidates turn_actions published to RabbitMQ.
  """
  use BattleRealTime.Contracts.Contract

  @primary_key false
  embedded_schema do
    field(:battle_id, :string)
    field(:turn, :integer)
    field(:timestamp, :string)

    embeds_many :actions, ActionItem, primary_key: false do
      field(:action, :string)
      field(:player_id, :integer)
      field(:id, :integer)
      field(:pokemon_id, :integer)
      field(:pokemon_switched_id, :integer)
      field(:targets, {:array, :string})
    end
  end

  def changeset(struct, params) do
    struct
    |> cast(params, [:battle_id, :turn, :timestamp])
    |> validate_required([:battle_id, :turn])
    |> cast_embed(:actions, required: true, with: &action_item_changeset/2)
  end

  defp action_item_changeset(struct, params) do
    struct
    |> cast(params, [:action, :player_id, :id, :pokemon_id, :pokemon_switched_id, :targets])
    |> validate_required([:action, :player_id])
    |> validate_action_fields()
  end

  defp validate_action_fields(changeset) do
    action = get_field(changeset, :action)

    case action do
      "attack" ->
        changeset
        |> validate_required([:id, :pokemon_id, :targets])

      "attack_switch" ->
        changeset
        |> validate_required([:id, :pokemon_id, :pokemon_switched_id, :targets])

      "switch" ->
        changeset
        |> validate_required([:pokemon_id])

      _ ->
        add_error(changeset, :action, "must be 'attack', 'attack_switch', or 'switch'")
    end
  end
end
